odmantic/engine.py

```diff
@@ -511,28 +551,37 @@ async def find_one(
             return None
         return results[0]
 
+    async def _save_collection_updates(
+        self,
+        collection_updates: CollectionUpdatesType,
+        session: "AsyncIOMotorClientSession",
+    ) -> None:
+        # reverse so that the last collections added, the ones for sub-documents, are
+        # saved first
+        for collection_name, updates in reversed(list(collection_updates.items())):
+            collection = self._get_collection_from_name(collection_name)
+            update_operations = [update[0] for update in updates]
+            update_instances = [update[1] for update in updates]
+            await collection.bulk_write(
+                update_operations,
+                session=session,
+            )
+            for inst in update_instances:
+                object.__setattr__(inst, "__fields_modified__", set())
+
     async def _save(
         self, instance: ModelType, session: "AsyncIOMotorClientSession"
     ) -> ModelType:
         """Perform an atomic save operation in the specified session"""
-        for ref_field_name in instance.__references__:
-            sub_instance = cast(Model, getattr(instance, ref_field_name))
-            await self._save(sub_instance, session)
-
-        fields_to_update = instance.__fields_modified__ | instance.__mutable_fields__
-        if len(fields_to_update) > 0:
-            doc = instance.doc(include=fields_to_update)
-            collection = self.get_collection(type(instance))
-            try:
-                await collection.update_one(
-                    instance.doc(include={instance.__primary_field__}),
-                    {"$set": doc},
-                    upsert=True,
-                    session=session,
-                )
-            except pymongo.errors.DuplicateKeyError as e:
+        collection_updates = self._prepare_document_updates(instance)
+        try:
+            await self._save_collection_updates(
+                collection_updates=collection_updates, session=session
+            )
+        except pymongo.errors.BulkWriteError as e:
+            if e.details["writeErrors"][0]["code"] == 11000:
```

Comment:
Maybe we could add a reference to the error code list in a comment.
The only thing I found is this https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml
But i think it would be helpful if something goes wrong.

Reply:
Cool! I added it. Indeed, there are practically no docs about the possible errors and when they are thrown. 😢